### PR TITLE
[engine] Track local active contract set

### PR DESCRIPTION
**Pull request by nickchapman-da**
Track locally created contracts through rollback, using new fields:
- `State.locallyCreated` and
- `ActiveLedgerState.locallyCreatedThisTimeline`

Address issues: https://github.com/digital-asset/daml/issues/14148 https://github.com/digital-asset/daml/issues/13835

TODO:
- ~~Adapt existing testcases in `ContractStateMachineSpec.scala`~~
- ~~New testcase (`rbCreate`) in `ContractStateMachineSpec.scala`~~
- ~~New testcase ("creates in try are rolled back") in `Engintest.scala`~~
- ~~Fix evaluation order tests~~
- ~~Improve new test in `EngineTest.scala` to specify the expected error (in `Left`)~~
- Add test running over ledger-API (in `ExceptionsIT.scala`)

### CODE REVIEW

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala [hunk 1]
```diff
@@ -1997,6 +1998,17 @@ class EngineTest
       )
       run(command) shouldBe a[Right[_, _]]
     }
+    "creates in try are rolled back" in {
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:2001]:
```suggestion
    "creates in rollback are rolled back" in {
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -1067,6 +1067,9 @@ private[lf] object SBuiltin {
             .foreach(nid =>
               throw SErrorDamlException(IE.ContractNotActive(coid, cached.templateId, nid))
             )
+          if (!onLedger.ptx.isActive(coid)) {
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1070]: shouldn’t we be able to drop the `onLedger.ptx.consumedBy` above? This looks like we’re now doing two activeness check.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 3]
```diff
@@ -89,11 +89,22 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
     *                      try blocks (interpretation) or Rollback nodes (iteration).
     */
   case class State private (
+      locallyCreated: Set[ContractId],
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:92]: Can we drop the `localContracts` Set from PartialTransaction so we don’t track this twice? I think a comment here would also be useful.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 4]
```diff
@@ -188,7 +207,10 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
         consuming: Boolean,
     ): State = {
       if (consuming) {
-        val consumedState = activeState.consume(targetId, nodeId)
+
+        val consumedState =
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:211]: why can we not use the `.consume` function anymore?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 5]
```diff
@@ -487,15 +512,14 @@ object ContractStateMachine {
     *    all operations involving a contract with a key update this map.
     */
   final case class ActiveLedgerState[+Nid](
+      locallyCreatedThisTimeline: Set[ContractId],
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:515]: please add a comment explaining what exactly this tracks and how it is different from locallyCreated.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/ContractStateMachineSpec.scala [hunk 6]
```diff
@@ -177,7 +177,7 @@ class ContractStateMachineSpec extends AnyWordSpec with Matchers with TableDrive
     val tx = builder.build()
     val expected = Right(
       Map(gkey("key1") -> Transaction.KeyActive(cid(1))) ->
-        ActiveLedgerState(Map(cid(0) -> (), cid(1) -> ()), Map(gkey("key1") -> KeyInactive))
+        ActiveLedgerState(Set(), Map(cid(0) -> (), cid(1) -> ()), Map(gkey("key1") -> KeyInactive))
```
Comment (cocreature) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/ContractStateMachineSpec.scala:180]:
```suggestion
        ActiveLedgerState(Set.empty, Map(cid(0) -> (), cid(1) -> ()), Map(gkey("key1") -> KeyInactive))
```

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/ContractStateMachineSpec.scala [hunk 7]
```diff
@@ -208,7 +208,7 @@ class ContractStateMachineSpec extends AnyWordSpec with Matchers with TableDrive
     val tx = builder.build()
     val expected = Right(
       Map(gkey("key1") -> Transaction.KeyActive(1), gkey("key2") -> Transaction.KeyActive(2)) ->
-        ActiveLedgerState(Map(cid(1) -> ()), Map(gkey("key1") -> KeyInactive))
+        ActiveLedgerState(Set(), Map(cid(1) -> ()), Map(gkey("key1") -> KeyInactive))
```
Comment (cocreature) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/ContractStateMachineSpec.scala:211]:
```suggestion
        ActiveLedgerState(Set.empty, Map(cid(1) -> ()), Map(gkey("key1") -> KeyInactive))
```

Comment (cocreature) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/ContractStateMachineSpec.scala:211]: same below

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1070]: You get a more precise error message from the original check.

Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1070]: Can we make isActive return an `Either[NodeId, ()]` or something like that if it fails due to `consumedBy` so we get the more precise error in that case? I don’t think we’re doing us any favor by keeping two sets of check.

Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: ok

**Comment by nickchapman-da** [comment]
all review comments addressed

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 8]
```diff
@@ -1062,11 +1062,13 @@ private[lf] object SBuiltin {
...
         case Some(cached) =>
-          onLedger.ptx.consumedBy
-            .get(coid)
-            .foreach(nid =>
+          onLedger.ptx.consumedByOrInactive(coid) match {
+            case Some(Left(nid)) =>
               throw SErrorDamlException(IE.ContractNotActive(coid, cached.templateId, nid))
-            )
+            case Some(Right(())) =>
+              throw SErrorDamlException(IE.ContractNotFound(coid))
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1069]: Let’s leave it for this PR but I think this maybe should throw ContractNotActive and we should make the node id optional. ContractNotFound is if the contract is not known to the ledger or speedy. Here we know about the contract but the create was rolled back.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 9]
```diff
@@ -225,13 +224,14 @@ private[speedy] case class PartialTransaction(
...
     aborted: Option[Tx.TransactionError],
     contractState: ContractStateMachine[NodeId]#State,
-    localContracts: Set[Value.ContractId],
     actionNodeLocations: BackStack[Option[Location]],
 ) {
 
   import PartialTransaction._
 
-  def consumedBy: Map[Value.ContractId, NodeId] = contractState.activeState.consumedBy
+  def consumedByOrInactive(cid: Value.ContractId): Option[Either[NodeId, Unit]] = {
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:232]: probably worth introducing a proper trait for this but let’s not do it in this pr.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 10]
```diff
@@ -89,11 +92,24 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
     *                      try blocks (interpretation) or Rollback nodes (iteration).
     */
   case class State private (
+      locallyCreated: Set[ContractId],
       globalKeyInputs: Map[GlobalKey, KeyInput],
       activeState: ActiveLedgerState[Nid],
       rollbackStack: List[ActiveLedgerState[Nid]],
   ) {
 
+    def consumedByOrInactive(cid: Value.ContractId): Option[Either[Nid, Unit]] = {
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:101]: please add a docs comment on this function that explains the return type.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 11]
```diff
@@ -69,6 +69,9 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
     *   keyset are in [[activeState]].[[ActiveLedgerState.keys]],
     *   and similarly for all [[ActiveLedgerState]]s in [[rollbackStack]].
     *
+    * @param locallyCreated
+    *   Tracks contracts created by this transaction. Even if subsequently rolled-back.
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:73]:
```suggestion
    *   Tracks all contracts created by this transaction including those under a rollback
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 12]
```diff
@@ -460,6 +487,9 @@ object ContractStateMachine {
   val KeyActive = Some
 
   /** Summarizes the updates to the current ledger state by nodes up to now.
+    *
+    * @param locallyCreatedThisTimeline
+    *   Tracks contracts created by this transaction. And which have not been rolled-back.
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:492]:
```suggestion
    *   Tracks contracts created by this transaction that have not been rolled back. This is a subset of `locallyCreated`.
```

**Review by cocreature — APPROVED** [review]
Thanks!

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala [hunk 13]
```diff
@@ -1997,6 +1998,17 @@ class EngineTest
       )
       run(command) shouldBe a[Right[_, _]]
     }
+    "creates in rollback are rolled back" in {
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:2001]:
```suggestion
    // TEST_EVIDENCE: Semantics: Rollback creates cannot be exercise
    "creates in rollback are rolled back" in {
```

**Comment by remyhaemmerle-da** [comment]
cc @matthiasS-da

**Comment by remyhaemmerle-da** [comment]
cc @andreaslochbihler-da

**Pull request merged**